src/Features/Core/Portable/ForeachToFor/AbstractForeachToForCodeRefactoringProvider.cs

```diff
@@ -0,0 +1,346 @@
+                if (countImpl == null || indexerImpl == null)
+                {
+                    continue;
+                }
+
+                if (!semanticFact.IsAssignableTo(indexerImpl.ReturnType, foreachType, model.Compilation))
+                {
+                    continue;
+                }
+
+                // implicitly implemented!
+                if (countImpl.ExplicitInterfaceImplementations.IsEmpty &&
+                    indexerImpl.ExplicitInterfaceImplementations.IsEmpty)
+                {
+                    explicitCast = null;
+                    countName = Count;
+                    return;
+                }
+
+                if (explicitInterface == null)
+                {
+                    explicitInterface = current.ToDisplayString();
+                }
+            }
+
+            // okay, we don't have implicitly implemented one, but we do have explicitly implemented one
+            if (explicitInterface != null)
+            {
+                explicitCast = explicitInterface;
+                countName = Count;
+            }
+        }
+
+        private static bool IsNullOrErrorType(ITypeSymbol type)
+        {
+            return type == null || type is IErrorTypeSymbol;
+        }
+
+        private static IMethodSymbol GetInterfaceMembers(ITypeSymbol interfaceType, string memberName)
+        {
+            var members = interfaceType.GetMembers(memberName);
+            if (!members.IsEmpty)
+            {
+                return (IMethodSymbol)members[0];
+            }
+
+            foreach (var current in interfaceType.AllInterfaces)
+            {
+                var member = GetInterfaceMembers(current, memberName);
+                if (member != null)
+                {
+                    return member;
+                }
+            }
+
+            return null;
+        }
+
+        private static bool CheckRequireCollectionStatement(IOperation operation)
+        {
+            return operation.Kind != OperationKind.LocalReference && operation.Kind != OperationKind.FieldReference;
+        }
+
+        private IOperation RemoveImplicitConversion(IOperation collection)
+        {
+            if (collection is IConversionOperation conversion && conversion.IsImplicit)
+            {
+                return RemoveImplicitConversion(conversion.Operand);
+            }
+
+            return collection;
+        }
+
+        private bool CheckForeachVariable(SemanticModel semanticModel, ISymbol foreachVariable, SyntaxNode foreachStatement)
+        {
+            var (start, end) = GetForeachBody(foreachStatement);
+            var dataFlow = semanticModel.AnalyzeDataFlow(start, end);
+
+            if (!dataFlow.Succeeded)
+            {
+                // if we can't get good analysis, assume it is written
+                return false;
+            }
+
+            return !dataFlow.WrittenInside.Contains(foreachVariable);
```

Comment:
how can a foreach variable be written to?

Reply:
it is an error in foreach. you can't write to foreach variable. but not an error in for loop (at least the way I converted it)

so, if foreach had an error and refactor to for loop, it suddenly become no error (semantic changes), so, this look for such case and not offer refactoring in such case.

Reply:
actually, I'm not sure if that's a good enough reason to disallow this. This would only fix code was previously an error, and this is just a refactoring, not an analyzer, so it wouldn't be obtrusive. I can imagine this scenario: what if someone is trying to modify the iteration variable, then realizes that's not possible and therefore wants to convert it to a for loop? This would turn an error into correct code but not in a way that's unexpected (I don't see any other interpretation of how that code could have been corrected). Many refactorings do this already.

Reply:
we can remove this restriction later. I worry people without realizing changing meaning of code.

Reply:
I'm with @Neme12  on this.  The code was in error before.  I think it's fine to allow the refactoring.  If you're concerned about them not realizing this, you can add a ConflictAnnotation at the places where the variable is written to.

Reply:
we can get feedback to see what people wants. while I was coding, I encountered this and I see benefit on not offering refactoring here.

Reply:
There is no feedback that can be provided here.  We've already seen thsi over and over again.  People simply assume that the product is busted when they do not get features but cannot tell why.  If they've written into their foreach-variable there is no reason that that should prevent htem from being able to convert to a for.  

Can you please clarify why you think it's bad to let someone do this?  It seems like we don't need the code *at all*. And, if we're going to check for this, because we do think it's important, we should at least do the right thing and add the conflict annotation rather than supressing the feature.

Reply:
I am not sure why adding warning is right thing? we can open an issue and talk about it and change if we want to. but not here.

Reply:
https://github.com/dotnet/roslyn/issues/25667